### DISCUSSION TITLE
Feature Request: Add filename and line index to CSV and JSON parsed rows (optionally)

### DISCUSSION THREAD

**Discussion by rustyconover** [duckdb/duckdb#9987] [Ideas]
When ingesting data from CSV or JSON it is quite helpful to have virtual columns like `line_index` or `filename` that contain the line index and filename where the row was produced.  This is very useful for data quality investigations where you want to work backwards to where the offending row originated.  `filename` is especially useful when a glob expression was used.

These `line_index` and `filename` columns could be added when the user opts in to having them produced.

`line_index` should count from all rows parsed from the file.

**Comment by Mytherin** [comment]
`filename` is already supported:

```sql
D copy (select 42 as i) to 'l1.csv';
D copy (select 84 as i) to 'l2.csv';
D select * from read_csv_auto('l*.csv', filename=true) order by filename;
┌───────┬──────────┐
│   i   │ filename │
│ int64 │ varchar  │
├───────┼──────────┤
│    42 │ l1.csv   │
│    84 │ l2.csv   │
└───────┴──────────┘

```

**Reply by adriangb** [reply]
Thank you for this as well, very nice touch:

```
D select filename, sum(row_count) as row_count from read_parquet('/**/*_stats.parquet', filename=true) group by filename order by row_count desc limit 10;
Binder Error:
Option filename adds column "filename", but a column with this name is also in the file. Try setting a different name: filename='<filename column name>'
```

**Comment by olsgaard** [comment]
Any plans on adding line number? I have ~100.000 csv files I'd like to import and the the line order is important to this data.

~Would partitioning by filename be able to guarantee me the correct line order?~ Partitioning by filename and adding row_number doesn't work, even if we preserve insertion order.

```sql
SET preserve_insertion_order = true;
SELECT 
  *, 
  ROW_NUMBER() OVER (PARTITION BY filename ORDER BY (SELECT 1)) AS line_number 
FROM
  read_csv_auto('l*.csv', filename=true) order by filename;
```

**Reply by Stene3** [reply]
I would be interested also in adding possibility to populate row index if possible. When importing huge number of rows without unique row ID we have to go over all results using a window function, but the row number over partition takes lots of resources when needs to run over list of millions of rows.

**Comment by olsgaard** [comment]
It does look like column order is preserved across files, so something like the following appears to work, but is it guarenteed?

```sql
WITH tbl1 AS (
  SELECT 
    *, 
    ROW_NUMBER() OVER (ORDER BY (SELECT 1)) AS line_number_across_files
  FROM
    read_csv_auto('l*.csv', filename=true) order by filename
)
SELECT
  *
  line_number_across_files - MIN(line_number_across_all) OVER (PARTITION BY filename) as line_number
FROM
  tbl1
 ```

It is more than a bit cumbersome and not obvious at all. I'm not sure why I need to use CTE, but if I move the `line_number` statement into the first `select`-statement, then I get a binder error:

```
Binder Error: window function calls cannot be nested 
```
